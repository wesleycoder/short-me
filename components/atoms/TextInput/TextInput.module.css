.textInput {
  border-radius: 15px 0 0 15px;
  background-color: antiquewhite;
  outline: none;
  min-width: 360px;
}

.textInput:focus {
  outline: 1px solid burlywood;
}

.textInput.invalid {
  outline: 2px solid crimson;
  background-color: mistyrose;
}
