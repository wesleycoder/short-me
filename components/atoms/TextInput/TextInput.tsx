import debounce from "lodash/debounce";
import {
  ChangeEvent,
  ComponentProps,
  EventHandler,
  KeyboardEventHandler,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";

interface TextInputProps extends ComponentProps<"input"> {
  delay?: number;
  defaultValue?: string;
  validate?: (value: string) => boolean;
  onChangeText: (value: string) => any | void;
  onValidate: (valid: boolean) => any | void;
}

type InputChangeEvent = ChangeEvent<HTMLInputElement>;

const useDebouncedOnChange = (
  onChange: EventHandler<InputChangeEvent>,
  debounceTime: number = 300,
  onAllChanges: EventHandler<InputChangeEvent> = () => {},
) => {
  const handler = useMemo(() =>
    debounce(onChange, debounceTime, {
      trailing: true,
    }), [
    debounceTime,
    onChange,
  ]);

  const handleAll = useCallback(
    (event: InputChangeEvent) => {
      handler(event);
      onAllChanges(event);
    },
    [handler, onAllChanges],
  );

  useEffect(() => () => handler.cancel(), [handler]);

  return handleAll;
};

export function TextInput(
  {
    type = "text",
    defaultValue = "",
    delay = 300,
    validate = () => true,
    onChangeText = () => {},
    onValidate = () => {},
    autoFocus = false,
    ...props
  }: TextInputProps,
) {
  const inputRef = useRef<HTMLInputElement>(null);
  const [isValid, setIsValid] = useState(
    () => (defaultValue ? validate(defaultValue) : true)
  );

  const onChange: EventHandler<InputChangeEvent> = useCallback(
    ({ target: { value } }) => onChangeText(value),
    [onChangeText],
  );

  const onKeyPress: KeyboardEventHandler<HTMLInputElement> = useCallback(
    (event) => {
      if (event.key === "Enter" && !isValid) {
        event.preventDefault();
      }
    },
    [isValid],
  );

  useEffect(
    () => {
      if (autoFocus) {
        inputRef.current?.focus();
      }
    },
    [autoFocus]
  )

  const onAllChanges: EventHandler<InputChangeEvent> = useCallback(
    ({ target: { value } }: InputChangeEvent) => {
      const valid = validate(value);

      if (isValid !== valid) {
        onValidate(valid);
      }

      setIsValid(valid);
    },
    [isValid, onValidate, validate],
  );

  const debouncedHandler = useDebouncedOnChange(onChange, delay, onAllChanges);

  return (
    <input
      {...props}
      ref={inputRef}
      autoFocus={autoFocus}
      formNoValidate
      type={type}
      defaultValue={defaultValue}
      onChange={debouncedHandler}
      onKeyPress={onKeyPress}
    />
  );
}
