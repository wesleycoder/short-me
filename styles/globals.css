html,
body {
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
  margin: 0;
  padding: 0;
  min-width: 480px;
}

html {
  min-height: 100vh;
}

body {
  display: flex;
  width: 100%;
  min-height: 100vh;
}

a {
  color: inherit;
  text-decoration: none;
}

* {
  box-sizing: border-box;
}

#__next {
  min-height: 100vh;
  width: 100%;
  background-color: slategray;
}
