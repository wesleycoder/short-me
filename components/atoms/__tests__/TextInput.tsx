import React from "react";
import { describe, it, expect } from "@jest/globals";
import { render, fireEvent } from "@testing-library/react";
import { TextInput } from "../TextInput";

describe("TextInput", () => {
  it("should render", () => {
    const label = "Text input";
    const { getByLabelText } = render(
      <TextInput aria-label={label} value={""} />
    );
    const input = getByLabelText(label);
    expect(input).toBeTruthy();
    expect(input).toHaveProperty("value", "");
    expect(input).toMatchSnapshot();
  });

  it("should render with initial text", () => {
    const label = "Text input";
    const value = "Initial text";
    const { getByLabelText } = render(
      <TextInput aria-label={label} value={value} />
    );
    const input = getByLabelText(label);
    expect(input).toBeTruthy();
    expect(input).toHaveProperty("value", value);
    expect(input).toMatchSnapshot();
  });
});
