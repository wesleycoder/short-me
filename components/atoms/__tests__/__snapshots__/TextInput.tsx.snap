// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should render 1`] = `
<input
  aria-label="Text input"
  formnovalidate=""
  type="text"
  value=""
/>
`;

exports[`TextInput should render with initial text 1`] = `
<input
  aria-label="Text input"
  formnovalidate=""
  type="text"
  value="Initial text"
/>
`;
